From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: kernel: improve spectre mitigation
Patch-mainline: no, SLES 12+15 only
References: bnc#1106934, LTC#171029

Description:  kernel: improve spectre mitigation
Symptom:      -
Problem:      The upstream kernel has a few more spectre related updates
              and fixes which are relevant for the various distribution
              targets as well. The most notable improvement is the
              etoken support.
Solution:     Backport upstream patches
Reproduction: -

Upstream-Description:
The patch is not upstream relevant, the description for older releases:

              s390: use expoline thunks for all branches generated by the BPF JIT

              git commit e1cf4befa297b149149f633eff746593e400c030
              "bpf, s390x: remove ld_abs/ld_ind"
              removed the code that generated the indirect branch "basr %b5,%w1"
              from the BPF JIT. Older versions of the BPF which still have support
              for LD_ABS/LD_IND need a patch to add the execute trampoline for
              this branch instruction.


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/net/bpf_jit_comp.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- a/arch/s390/net/bpf_jit_comp.c
+++ b/arch/s390/net/bpf_jit_comp.c
@@ -1275,8 +1275,13 @@ call_fn:
 			/* agfr %b2,%src (%src is s32 here) */
 			EMIT4(0xb9180000, BPF_REG_2, src_reg);
 
-		/* basr %b5,%w1 (%b5 is call saved) */
-		EMIT2(0x0d00, BPF_REG_5, REG_W1);
+		if (IS_ENABLED(CC_USING_EXPOLINE) && !nospec_disable) {
+			/* brasl %r5,__s390_indirect_jump_r1 */
+			EMIT6_PCREL_RILB(0xc0050000, BPF_REG_5, jit->r1_thunk_ip);
+		} else {
+			/* basr %b5,%w1 (%b5 is call saved) */
+			EMIT2(0x0d00, BPF_REG_5, REG_W1);
+		}
 
 		/*
 		 * Note: For fast access we jump directly after the
